libraries/chain/controller.cpp

```diff
@@ -1512,27 +1512,39 @@ struct controller_impl {
                // irreversible. Instead, this moves irreversible as much as possible and allows the next maybe_switch_forks call to apply these
                // non-validated blocks. After the maybe_switch_forks call (before next produced block or on next received block), irreversible
                // can then move forward on the then validated blocks.
-               return read_mode == db_read_mode::IRREVERSIBLE || bsp->is_valid();
+               // In irreversible mode, break every ~500ms to allow other tasks (e.g. get_info, SHiP) opportunity to run. There is a post
+               // for every incoming blocks; enough posted tasks to apply all blocks queued to the fork db.
+               if (irreversible_mode()) {
+                  if (!replaying && fc::time_point::now() - start > fc::milliseconds(500))
+                     return false;
+                  return true;
+               }
+               return bsp->is_valid();
             };
 
-            std::vector<std::future<std::vector<char>>> v;
-            v.reserve( branch.size() );
-            for( auto bitr = branch.rbegin(); bitr != branch.rend() && should_process(*bitr); ++bitr ) {
-
-               v.emplace_back( post_async_task( thread_pool.get_executor(), [b=(*bitr)->block]() { return fc::raw::pack(*b); } ) );
+            using packed_block_future = std::future<std::vector<char>>;
+            std::vector<packed_block_future> v;
+            if (!irreversible_mode()) {
+               v.reserve( branch.size() );
+               for( auto bitr = branch.rbegin(); bitr != branch.rend() && should_process(*bitr); ++bitr ) {
+                  v.emplace_back( post_async_task( thread_pool.get_executor(), [b=(*bitr)->block]() { return fc::raw::pack(*b); } ) );
+               }
             }
             auto it = v.begin();
 
             for( auto bitr = branch.rbegin(); bitr != branch.rend() && should_process(*bitr); ++bitr ) {
-               if (!apply_irreversible_block(forkdb, *bitr))
-                  break;
+               packed_block_future f;
+               if (irreversible_mode()) {
+                  f = post_async_task( thread_pool.get_executor(), [b=(*bitr)->block]() { return fc::raw::pack(*b); } );
+                  if (!apply_irreversible_block(forkdb, *bitr))
+                     break;
+               }
 
                emit( irreversible_block, std::tie((*bitr)->block, (*bitr)->id()), __FILE__, __LINE__ );
 
                // blog.append could fail due to failures like running out of space.
                // Do it before commit so that in case it throws, DB can be rolled back.
-               blog.append( (*bitr)->block, (*bitr)->id(), it->get() );
-               ++it;
+               blog.append( (*bitr)->block, (*bitr)->id(), irreversible_mode() ? f.get() : it++->get() );
```

Comment:
Hum, I think the issue could still occur. 

Even in non-irreversible mode (where we queue all the `pack` tasks), the rest of the second loop (`blog.append()` for example) could take some time (suppose the system is swapping because the memory is over-committed) causing us to exit the second loop leaving many of the `pack` tasks queued, only to be re-enqueued immediately.

Reply:
It will never exit the loop early in non-irreversible mode. There is no mechanism to interrupt the loop in non-irreversible mode.

Reply:
Oh yes, makes sense, I had missed that.
Is it possible still that we would get an exception? Maybe we should also add an `elog` before the throw at line `1567`?

Reply:
yes, we should log that exception. I'll add that.

Reply:
I honestly question whether the optimization of posting the `pack` to a thread pool is worth it, considering it makes the code harder to understand.

Reply:
Certainly can be removed if/when #1062 is worked. When it was added it did make a measurable difference. That was with legacy consensus which typically moves LIB by 12 instead of 2. Regardless, seems like it would be useful during syncing until #1062 is complete.